src/sentry/similarity.py

```diff
@@ -0,0 +1,212 @@
+    Convert a mapping of distinct quantities to a mapping of proportions of the
+    total quantity.
+    """
+    total = float(sum(value.values()))
+    return {k: (v / total) for k, v in value.items()}
+
+
+def get_distance(target, other):
+    """\
+    Calculate the N-dimensional Euclidian between two mappings.
+
+    The mappings are used to represent sparse arrays -- if a key is not present
+    in both mappings, it's assumed to be 0 in the mapping where it is missing.
+    """
+    return math.sqrt(
+        sum(
+            (target.get(k, 0) - other.get(k, 0)) ** 2
+            for k in set(target) | set(other)
+        )
+    )
+
+
+def get_similarity(target, other):
+    assert len(target) == len(other)
+    return 1 - sum(
+        map(
+            lambda (left, right): get_distance(
+                left,
+                right,
+            ) / math.sqrt(2),
+            zip(target, other)
+        )
+    ) / len(target)
+
+
+class MinHashIndex(object):
+    """
+    Implements an index that can be used to efficiently search for items that
+    share similar characteristics.
+
+    This implementation is based on MinHash (which is used quickly identify
+    similar items and estimate the Jaccard similarity of their characteristic
+    sets) but this implementation extends the typical design to add the ability
+    to record items by an arbitrary key. This allows querying for similar
+    groups that contain many different characteristic sets.
+
+    The ``rows`` parameter is the size of the hash ring used to collapse the
+    domain of all tokens to a fixed-size range. The total size of the LSH
+    signature is ``bands * buckets``. These attributes control the distribution
+    of data within the index, and modifying them after data has already been
+    written will cause data loss and/or corruption.
+
+    This is modeled as two data structures:
+
+    - A bucket frequency sorted set, which maintains a count of what buckets
+      have been recorded -- and how often -- in a ``(band, key)`` pair. This
+      data can be used to identify what buckets a key is a member of, and also
+      used to identify the degree of bucket similarity when comparing with data
+      associated with another key.
+    - A bucket membership set, which maintains a record of what keys have been
+      record in a ``(band, bucket)`` pair. This data can be used to identify
+      what other keys may be similar to the lookup key (but not the degree of
+      similarity.)
+    """
+    BUCKET_MEMBERSHIP = '0'
+    BUCKET_FREQUENCY = '1'
+
+    def __init__(self, cluster, rows, bands, buckets):
+        self.namespace = b'sim'
+
+        self.cluster = cluster
+        self.rows = rows
+
+        generator = random.Random(0)
+
+        def shuffle(value):
+            generator.shuffle(value)
+            return value
+
+        self.bands = [
+            [shuffle(range(rows)) for _ in xrange(buckets)]
+            for _ in xrange(bands)
+        ]
```

Comment:
I'm generally against logic in the constructor like this, but it also felt weird to make the caller pass in the permutations instead of declaring the band and bucket counts. (Also would still love to figure out the independent hash functions stuff but not holding my breath.)